### PR TITLE
Add Imperva WAF integration

### PR DESCRIPTION
**Pull request by adi88d**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/17696

## Description
Added integration for Imperva WAF.

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

**Comment by adi88d** [comment]
@kirbles19  yaml & docs are ready to be reviewed.

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+    site = args.get('site-name-to-apply')
+    server_group = args.get('server-group-name-to-apply')
+    web_service = args.get('web-service-name-to-apply')
+    match_criteria_json = args.get('match-criteria-json')
+
+    body = generate_policy_data_body(args)
+
+    if match_criteria_json:
+        try:
+            match_criteria_json = json.loads(match_criteria_json)
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:350]: same here

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+    entry_context = {f'{INTEGRATION_CONTEXT_NAME}.IpGroup.Name': group_name}
+    return_outputs(f'Group {group_name} created successfully', entry_context, raw_res)
+
+
+@logger
+def update_ip_group_command(client, args):
+    group_name = args.get('group-name')
+    body = generate_ip_groups_entries(args)
+    raw_res = client.do_request('PUT', f'conf/ipGroups/{group_name}/data', json_data=body)
+    return_outputs(f'Group {group_name} updated successfully', {}, raw_res)
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:328]: same here

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+
+@logger
+def sites_list_command(client, args):
+    raw_res = client.do_request('GET', 'conf/sites')
+    sites = []
+    if raw_res.get('sites'):
+        sites = raw_res['sites']
+        for i, element in enumerate(sites):
+            sites[i] = {'Name': sites[i]}
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:205]:
```suggestion
    sites = [{'Name': site} for site in raw_res.get('sites', [])]
```

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+
+def generate_policy_data_body(args):
+    severity = args.get('severity')
+    action = args.get('action')
+    followed_action = args.get('followed-action')
+
+    body = {}
+
+    if args.get('enabled'):
+        body['enabled'] = True if args['enabled'] == 'True' else False
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:65]: same for all
```suggestion
        body['enabled'] = args['enabled'] == 'True'
```

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+
+import json
+import requests
+import traceback
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+''' CONSTANTS '''
+INTEGRATION_CONTEXT_NAME = 'ImpervaWaf'
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:14]: why not `ImpervaWAF`?

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.yml [hunk 6]
```diff
@@ -250,8 +250,20 @@ script:
     execution: false
     name: imperva-waf-ip-group-create
     outputs:
-    - contextPath: ImpervaWaf.IpGroup.Name
-      description: The name of the IP Group
+    - contextPath: ImpervaWAF.IpGroup.Entries.Type
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.yml:253]: shouldn't this include the group name too?

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.yml [hunk 7]
```diff
@@ -315,6 +327,22 @@ script:
     description: Add or remove rows in an IP Group indicated by ip Group Name.
     execution: false
     name: imperva-waf-ip-group-update-entries
+    outputs:
+    - contextPath: ImpervaWAF.IpGroup.Entries.Type
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.yml:331]: same here

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 8]
```diff
@@ -45,6 +45,74 @@ def login(self):
         extract_errors(res)
         self.session_id = res.get('session-id')
 
+    def get_ip_group_entities(self, group_name, table_name):
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:48]: `get` function that doesn't return anything? please update the name or return the triplet (hr, context, raw) to the calling function and call `return_outputs` there

File: Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py [hunk 9]
```diff
@@ -45,6 +45,74 @@ def login(self):
...
+                            'CidrMask': entry.get('cidrMask')})
+
+        human_readable = tableToMarkdown(table_name, entries, removeNull=True,
+                                         headers=['Type', 'IpAddressFrom', 'IpAddressTo', 'NetworkAddress', 'CidrMask'])
+        entry_context = {f'{INTEGRATION_CONTEXT_NAME}.IpGroup(val.Name===obj.Name)':
+                         {'Name': group_name, 'Entries': entries}}
+
+        return_outputs(human_readable, entry_context, raw_res)
+
+    def get_custom_policy_outputs(self, policy_name, table_name):
```
Comment (yaakovi) [inline Packs/Imperva_WAF/Integrations/ImpervaWAF/ImpervaWAF.py:65]: same

**Comment by kirbles19** [comment]
@adi88d Done.

**Pull request merged**